Documentation/design-docs/arm64-intrinsics.md

```diff
@@ -0,0 +1,405 @@
+`ARM64` platforms.  This set will include general purpose instructions.  Investigation is needed to determine if
+any of these need intrinsics or whether this is an empty set.
+
+As further extensions are released, this set of intrinsics will grow.
+
+### Intrinsic Method Names
+
+Intrinsics will be named to describe functionality.  Names will not correspond to specific named
+assembly instructions.
+
+Where precedence exists within the `System.Runtime.Intrinsics.X86` namespace, identical method names will be
+chosen: `Add`, `Multiply`, `Load`, `Store` ...
+
+It is also worth noting `System.Runtime.Intrinsics.X86` naming conventions will include the suffix `Scalar` for
+operations which take vector argument(s), but contain an implicit cast(s) to the base type and therefore operate only
+on the first item of the argument vector(s).
+
+### Intinsic Method Argument Types
+
+Intrinsic methods will typically use a standard set of argument types:
++ Integer type: `byte`, `sbyte`, `short`, `ushort`, `int`, `uint`, `long`, `ulong`
++ Floating types: `double`, `single`, `System.Half`
++ Vector types: Vector128<T>
++ SVE will add new vector types: TBD
+
+## Intrinsic Interface Documentation
+
++ Namespace
++ Each `static class` will
+  + Briefly document corresponding `System Register Field and Value` from ARM specification.
+  + Document use of IsSupported property
+  + Optionally summarize set of methods enabled by the extension
++ Each intrinsic method will
+  + Document underlying `ARM64` assembly instruction
+  + Optionally, briefly summarize operation performed
+    + In many cases this may be unnecessary: `Add`, `Multiply`, `Load`, `Store`
+    + In some cases this may be difficult to do correctly. (Crypto instructions)
+  + Optionally mention corresponding compiler gcc, clang, and/or MSVC intrinsics
+    + Review of existing documentation shows `ARM64` intrinsics are mostly absent or undocumented so
+    initially this will not be necessary for `ARM64`
+    + See gcc manual "AArch64 Built-in Functions"
+    + MSVC ARM64 documentation has not been publically released
+
+## Phased Implementation
+
+### Implementation Priorities
+
+As rough guidelines for order of implementation:
+
++ Baseline functionality will be prioritized over architectural extensions
++ Architectural extensions will typically be prioritized in age order.  Earlier extensions will be added first
+  + This is primarily driven by availability of hardware.  Features released in earlier will be prevalent in
+  more hardware.
++ Priorities will be driven by optimization efforts and requests
+  + Priority will be given to intrinsics which are equivalent/similar to those actively used in libraries for other platforms
+  + Priority will be given to intrinsics which have already been implemented for other platforms
+
+### API review
+
+Intrinsics will extend the API of CoreCLR.  They will need to follow standard API review practices.
```

Comment:
It was proposed for Intel HW intrinsics to be a part of .NET Standard API. Perhaps it should be made clear that the Arm HW intrinsics could be a part of .NEt Standard as well. see https://github.com/dotnet/corefx/pull/24490 https://github.com/dotnet/corefx/issues/24346